### ISSUE TITLE
Multiple Compile issues - compiles clean under PNut (files attached)

### ISSUE THREAD

**Issue by ironsheep** [totalspectrum/spin2cpp#120]
Platform Raspberry Pi
Issue: incompatibility with PNut compiler
Provided: I've included a tar file with the full source code and the full output log.

Multiple issues found in the code. Here is  the command used and partial output:
```bash
> Executing task: flexspin -2 -Wabs-paths -Wmax-errors=99 isp_measure_panel.spin2 <

Propeller Spin/PASM Compiler 'FlexSpin' (c) 2011-2021 Total Spectrum Software Inc.
Version 5.3.0-beta-v5.2.0-27-ga9b51089 Compiled on: Mar 19 2021
isp_measure_panel.spin2
|-jm_serial.spin2
|-|-jm_nstr.spin2
|-isp_hub75_color.spin2
|-|-isp_hub75_screenAccess.spin2
|-|-|-isp_hub75_hwGeometry.spin2
|-isp_hub75_screenUtils.spin2
|-|-isp_hub75_color.spin2
|-|-isp_hub75_screenAccess.spin2
|-isp_hub75_screenAccess.spin2
|-isp_hub75_display.spin2
|-|-isp_hub75_screenAccess.spin2
|-|-isp_hub75_screenUtils.spin2
|-|-isp_hub75_color.spin2
|-|-isp_hub75_panel.spin2
|-|-|-isp_hub75_screenAccess.spin2
|-|-|-isp_hub75_rgb3bit.spin2
|-|-|-|-isp_hub75_screenAccess.spin2
|-|-|-|-isp_hub75_color.spin2
/home/pi/Projects-extGit/P2-Panel-Light-Measure/src/isp_hub75_rgb3bit.spin2:124: error: syntax error, unexpected ELSEIF
/home/pi/Projects-extGit/P2-Panel-Light-Measure/src/isp_hub75_rgb3bit.spin2:126: error: syntax error, unexpected ELSEIF
/home/pi/Projects-extGit/P2-Panel-Light-Measure/src/isp_hub75_rgb3bit.spin2:129: error: syntax error, unexpected ELSEIF
|-|-|-isp_hub75_color.spin2
/home/pi/Projects-extGit/P2-Panel-Light-Measure/src/isp_hub75_panel.spin2:565: error: syntax error, unexpected ELSE
|-|-isp_hub75_fonts.spin2
|-|-isp_hub75_scrollingText.spin2
|-|-|-isp_hub75_screenAccess.spin2
|-|-|-isp_hub75_screenUtils.spin2
|-|-|-isp_hub75_fonts.spin2
|-isp_bh1750.spin2
|-|-jm_i2c.spin2
error: Internal error: using longjmp/setjmp but feature not marked
error: Internal error: using longjmp/setjmp but feature not marked
...
/home/pi/Projects-extGit/P2-Panel-Light-Measure/src/isp_hub75_panel.spin2:134: error: must use an immediate with labels in inline asm
...
/home/pi/Projects-extGit/P2-Panel-Light-Measure/src/isp_hub75_panel.spin2:292: error: inline assembly of this item not supported yet
...
/home/pi/Projects-extGit/P2-Panel-Light-Measure/src/isp_hub75_panel.spin2:449: error: inline assembly of this item not supported yet
error: Internal error: using longjmp/setjmp but feature not marked
...
/home/pi/Projects-extGit/P2-Panel-Light-Measure/src/isp_hub75_rgb3bit.spin2:570: error: inline assembly of this item not supported yet
The terminal process "/bin/bash '-c', 'flexspin -2 -Wabs-paths -Wmax-errors=99 isp_measure_panel.spin2'" terminated with exit code: 1.
```

View full output in `compile.log` found within tar.gz file.  Three files are marked with errors.  All compile clean using PNut v35L

Some of this appears to be interspersed debug() statements. But there are more issues.

[failed-source.tar.gz](https://github.com/totalspectrum/spin2cpp/files/6247010/failed-source.tar.gz)

**Comment by totalspectrum** [comment]
Thank you for the bug report. There were certainly some debug issues (the removal of debugs left a gap that caused indentation to be messed up in nested if/else), and I've fixed that. Unfortunately most of the errors are due to a fundamental difference between PNut and FlexSpin, which is that FlexSpin executes inline assembly from LUT or HUB, whereas PNut always executes it from COG. The FlexSpin design pre-dates PNut having inline assembly, and unfortunately the choice does mean that declaring variables with LONG is not supported in FlexSpin. I'll try to make this clearer in the documentation. 

So if you want your code to work with FlexSpin, instead of writing "columnIdx long 0" inside inline assembly you'll have to declare a local variable "columnIdx" in the enclosing function and initialize it to 0 (if the initial value matters).

**Comment by ironsheep** [comment]
Yes, I know about the local variables. With Chip's guidance, I specifically moved away from that solution because I needed more than 16 vars. and if I used local vars. I ended up having to re-use some of the variables which made the code much harder to understand.

Does this make sense?

**Comment by totalspectrum** [comment]
Perhaps Chip could modify his compiler to allow more than 16 local variables.

**Comment by ironsheep** [comment]
You might guess that I don't leave many avenues unexplored. ;-)

That was a good discussion and deliberation with Chip. In the end, 16 was the best tradeoff possible. Then given this declaring more as variables in the inline code, we don't have to face this 16 limitation in the exceptional cases while 16 is more than enough in most cases.

**Comment by totalspectrum** [comment]
16 may be a good tradeoff for Chip's compiler/interpreter, but it isn't universally. flexspin has to support BASIC and C, which both allow for unlimited local variables, so it needs a lot more space in COG memory for those variables (obviously it's not unlimited in practice, but it's more like 256 local variables than 16). Obviously it's possible in theory for Chip to modify his system to support more than 16 variables, but for whatever reason I guess he doesn't want to.

**Comment by totalspectrum** [comment]
Stephen, the version checked in to github now is able to compile your code. I don't know if it works correctly or not, I don't have the hardware to test it.

**Comment by ironsheep** [comment]
Eric, this is exciting news.  I'm glad to test and feedback what I find out. 
More soon...

I can't thank you enough for making this happen!

**Comment by ironsheep** [comment]
Eric,  I compiled and ran my code using your changes (github latest) and PNut. Two tests. Both are now functionally the same against the hardware.  I ran multiple bodies of code showing different displays and all are displaying correctly.

Good work!

I have to say this is now the first time I'm building and running my project from all three platforms directly (Raspberry Pi, Mac, and Windows machines), no moving files to another machine to compile and download. 

Thank you.
